Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverCoins/Coins.hs

```diff
@@ -0,0 +1,205 @@
+import qualified Storage.Queries.Coins.CoinHistory as CHistory
+import qualified Storage.Queries.Coins.CoinsConfig as DCQ
+import qualified Storage.Queries.Person as Person
+import Storage.Queries.Ride as Ride
+
+type EventFlow m r = (MonadFlow m, EsqDBFlow m r, CacheFlow m r, MonadReader r m, HasField "minTripDistanceForReferralCfg" r (Maybe HighPrecMeters))
+
+mkCoinAccumulationByDriverIdKey :: Id DP.Person -> Text -> Text
+mkCoinAccumulationByDriverIdKey driverId date = "DriverCoinBalance:DriverId:" <> driverId.getId <> ":" <> date
+
+getCoinAccumulationByDriverIdKey :: (MonadFlow m, EsqDBFlow m r, CacheFlow m r) => Id DP.Person -> Text -> m (Maybe Int)
+getCoinAccumulationByDriverIdKey driverId currentDate = Hedis.get (mkCoinAccumulationByDriverIdKey driverId currentDate)
+
+setCoinAccumulationByDriverIdKey :: (MonadFlow m, EsqDBFlow m r, CacheFlow m r) => Id DP.Person -> Text -> Int -> m ()
+setCoinAccumulationByDriverIdKey driverId currentDate count = do
+  void $ Hedis.incrby (mkCoinAccumulationByDriverIdKey driverId currentDate) (fromIntegral count)
+  Hedis.expire (mkCoinAccumulationByDriverIdKey driverId currentDate) 86400
+
+getCoinsByDriverId :: (MonadFlow m, EsqDBFlow m r, CacheFlow m r) => Id DP.Person -> Maybe Int -> m Int
+getCoinsByDriverId driverId mbCoinUpdateValue = do
+  now <- getCurrentTime
+  let currentDate = show $ utctDay now
+  coinKeyExists <- getCoinAccumulationByDriverIdKey driverId currentDate
+  case (coinKeyExists, mbCoinUpdateValue) of
+    (Just _, Just coinUpdateValue) -> do
+      void $ Hedis.incrby (mkCoinAccumulationByDriverIdKey driverId currentDate) (fromIntegral coinUpdateValue)
+      pure coinUpdateValue
+    (Just coinBalance, Nothing) -> pure coinBalance
+    (Nothing, _) -> do
+      totalCoins <- CHistory.getTotalCoins driverId
+      let coinBalance = sum $ map (\coinHistory -> coinHistory.coins - coinHistory.coinsUsed) totalCoins
+      Hedis.whenWithLockRedis (mkCoinAccumulationByDriverIdKey driverId currentDate) 60 $ do
+        setCoinAccumulationByDriverIdKey driverId currentDate coinBalance
+      pure coinBalance
+
+driverCoinsEvent :: EventFlow m r => Id DP.Person -> Id DM.Merchant -> Id DMOC.MerchantOperatingCity -> DCT.DriverCoinsEventType -> m ()
+driverCoinsEvent driverId merchantId merchantOpCityId eventType = do
+  logDebug $ "Driver Coins Event Triggered for merchantOpCityId - " <> merchantOpCityId.getId <> " and driverId - " <> driverId.getId
+  coinConfiguration <- DCQ.fetchFunctionsOnEventbasis eventType merchantId merchantOpCityId
+  finalCoinsValue <- sum <$> forM coinConfiguration (\cc -> calculateCoins eventType driverId merchantId merchantOpCityId cc.eventFunction cc.expirationAt cc.coins)
+  driver <- Person.findById driverId >>= fromMaybeM (PersonNotFound driverId.getId)
+  void $ Person.updateTotalEarnedCoins driverId (finalCoinsValue + driver.totalEarnedCoins)
+  void $ getCoinsByDriverId driverId (Just finalCoinsValue)
```

Comment:
create a separate function for incr and call that directly....this name doesn't make sense for a function which updates stuff, also this feature of this function is used only once, take that part out into another function.